embassy-usb-hid/src/lib.rs

```diff
@@ -0,0 +1,507 @@
+pub enum ReportId {
+    In(u8),
+    Out(u8),
+    Feature(u8),
+}
+
+impl ReportId {
+    fn try_from(value: u16) -> Result<Self, ()> {
+        match value >> 8 {
+            1 => Ok(ReportId::In(value as u8)),
+            2 => Ok(ReportId::Out(value as u8)),
+            3 => Ok(ReportId::Feature(value as u8)),
+            _ => Err(()),
+        }
+    }
+}
+
+pub struct State<'a, const IN_N: usize, const OUT_N: usize> {
+    control: MaybeUninit<Control<'a, OUT_N>>,
+    out_signal: Signal<(usize, [u8; OUT_N])>,
+}
+
+impl<'a, const IN_N: usize, const OUT_N: usize> State<'a, IN_N, OUT_N> {
+    pub fn new() -> Self {
+        State {
+            control: MaybeUninit::uninit(),
+            out_signal: Signal::new(),
+        }
+    }
+}
+
+pub struct HidClass<'d, D: Driver<'d>, const IN_N: usize, const OUT_N: usize> {
+    input: ReportWriter<'d, D, IN_N>,
+    output: ReportReader<'d, D, OUT_N>,
+}
+
+impl<'d, D: Driver<'d>, const IN_N: usize, const OUT_N: usize> HidClass<'d, D, IN_N, OUT_N> {
+    /// Creates a new HidClass.
+    ///
+    /// poll_ms configures how frequently the host should poll for reading/writing
+    /// HID reports. A lower value means better throughput & latency, at the expense
+    /// of CPU on the device & bandwidth on the bus. A value of 10 is reasonable for
+    /// high performance uses, and a value of 255 is good for best-effort usecases.
+    ///
+    /// This allocates two endpoints (IN and OUT).
+    /// See new_ep_in (IN endpoint only) and new_ep_out (OUT endpoint only) to only create a single
+    /// endpoint.
+    pub fn new(
+        builder: &mut UsbDeviceBuilder<'d, D>,
+        state: &'d mut State<'d, IN_N, OUT_N>,
+        report_descriptor: &'static [u8],
+        request_handler: Option<&'d dyn RequestHandler>,
+        poll_ms: u8,
+    ) -> Self {
+        let ep_out = Some(builder.alloc_interrupt_endpoint_out(64, poll_ms));
+        let ep_in = Some(builder.alloc_interrupt_endpoint_in(64, poll_ms));
+        Self::new_inner(
+            builder,
+            state,
+            report_descriptor,
+            request_handler,
+            ep_out,
+            ep_in,
+        )
+    }
+
+    /// Creates a new HidClass with the provided UsbBus & HID report descriptor.
+    /// See new() for more details.
+    pub fn new_ep_in(
+        builder: &mut UsbDeviceBuilder<'d, D>,
+        state: &'d mut State<'d, IN_N, OUT_N>,
+        report_descriptor: &'static [u8],
+        request_handler: Option<&'d dyn RequestHandler>,
+        poll_ms: u8,
+    ) -> Self {
+        let ep_out = None;
+        let ep_in = Some(builder.alloc_interrupt_endpoint_in(64, poll_ms));
+        Self::new_inner(
+            builder,
+            state,
+            report_descriptor,
+            request_handler,
+            ep_out,
+            ep_in,
+        )
+    }
+
+    /// Creates a new HidClass with the provided UsbBus & HID report descriptor.
+    /// See new() for more details.
+    pub fn new_ep_out(
```

Comment:
I think this is illegal? The [HID specs](https://www.usb.org/document-library/device-class-definition-hid-111), page 10 say the INTERRUPT IN endpoint is mandatory.

Reply:
Agreed, we should remove it.